app/routes.py

```diff
@@ -1,97 +1,232 @@
+        else:
+            return jsonify({"error": "Failed to add harvest job."}), 400
+    else:
+        return jsonify({"Please provide harvest job with json format."})
 
-@mod.route('/harvest_source/<source_id>', methods=['GET'])
-def get_harvest_source(source_id):
+@mod.route("/harvest_job/", methods=["GET"])    
+@mod.route("/harvest_job/<job_id>", methods=["GET"])
+def get_harvest_job(job_id=None):
     try:
-        result = db.get_harvest_source(source_id)
-        return result
+        if job_id:
+            job = db.get_harvest_job(job_id)
+            return jsonify(job) if job else ("Not Found", 404)
+
+        source_id = request.args.get("harvest_source_id")
+        if source_id:
+            job = db.get_harvest_job_by_source(source_id)
+            if not job:
+                return "No harvest jobs found for this harvest source", 404
+        else:
+            job = db.get_all_harvest_jobs()
+    
+        return jsonify(job)
     except Exception:
-        return " provide source_id"
+        return "Please provide correct job_id or harvest_source_id"
 
-@mod.route("/harvest_source/<string:source_id>", methods=["PUT"])
-def update_harvest_source(source_id):
-    result = db.update_harvest_source(source_id, request.json)
+@mod.route("/harvest_job/<job_id>", methods=["PUT"])
+def update_harvest_job(job_id):
+    result = db.update_harvest_job(job_id, request.json)
+    return result
+
+@mod.route("/harvest_job/<job_id>", methods=["DELETE"])
+def delete_harvest_job(job_id):
+    result = db.delete_harvest_job(job_id)
     return result
 
-@mod.route('/harvest_job/<job_id>', methods=['GET'])
-def get_harvest_job(job_id):
+@mod.route("/harvest_error/add", methods=["POST", "GET"])
+def add_harvest_error():
+    if request.is_json:
+        error = db.add_harvest_error(request.json)
+        if error:
+            return jsonify({"message": f"Added new harvest error with ID: {error.id}"})
+        else:
+            return jsonify({"error": "Failed to add harvest error."}), 400
+    else:
+        return jsonify({"Please provide harvest error with json format."})
+
+@mod.route("/harvest_error/", methods=["GET"])    
+@mod.route("/harvest_error/<error_id>", methods=["GET"])
+def get_harvest_error(error_id=None):
     try:
-        result = db.get_harvest_job(job_id)
-        return result
+        if error_id:
+            error = db.get_harvest_error(error_id)
+            return jsonify(error) if error else ("Not Found", 404)
+
+        job_id = request.args.get("harvest_job_id")
+        if job_id:
+            error = db.get_harvest_error_by_job(job_id)
+            if not error:
+                return "No harvest errors found for this harvest job", 404
+        else:
+            # for test, will remove later
+            error = db.get_all_harvest_errors()
+    
+        return jsonify(error)
     except Exception:
-        return "provide job_id"
+        return "Please provide correct error_id or harvest_job_id"
```

Comment:
remove `harvest_job_id`

Reply:
since we may get job either with error_id or job_id, the message is for both possible situations.